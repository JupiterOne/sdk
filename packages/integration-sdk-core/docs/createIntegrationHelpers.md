# createIntegrationHelpers

This function is used to create typeful integration helpers .

```typescript
function createIntegrationHelpers(
  options: CreateIntegrationHelpersOptions,
): IntegrationHelpers;
```

#### Parameters

- `options` - An object containing the following properties:
  - `integrationName`: Integration name. **_Example: 'aws', 'microsoft_365'_**
  - `classSchemaMap`: Typebox class schema map.

#### Returns

An object containing the following properties:

- `createEntityType`: A function to generate entity types with
  `${integrationName}_${entityName}` pattern.
- `createIntegrationEntity`: A function to create a entity metadata and a typed
  create entity function.

#### Example

```typescript
import { createIntegrationHelpers } from '@jupiterone/integration-sdk-core';
import { classSchemaTypeboxMap } from '@jupiterone/data-model';

const { createEntityType, createIntegrationEntity } = createIntegrationHelpers({
  integrationName: 'my_awesome_integration',
  schemaMap: classSchemaTypeboxMap,
});

const [USER_ENTITY, createUserAssignEntity] = createIntegrationEntity({
  resourceName: 'User',
  _class: ['User'],
  _type: createEntityType('user'), // This will generate "my_awesome_integration_user", but you are free to not use the createEntityType helper
  description: 'Entity description', // This will be used in the json schema
  schema: Type.Object({
    name: Type.String(),
  }),
});

// _type and _class will be generated automatically
createUserAssignEntity({
  _key: `${Entities.ACCOUNT._type}|${_integrationInstanceId}`,
  name,
});
```

# How to use it in a current graph integration

1. Create a file named helpers.ts in the src folder of your integration
2. Add the following code to the file:

```typescript
import { createIntegrationHelpers } from '@jupiterone/integration-sdk-core';
import { classSchemaTypeboxMap } from '@jupiterone/data-model';

export const { createEntityType, createIntegrationEntity } =
  createIntegrationHelpers({
    integrationName: INTEGRATION_NAME,
    schemaMap: classSchemaTypeboxMap,
  });
```

3. Replace INTEGRATION_NAME with the name of your integration
4. Create a file named entities.ts in the src folder of your integration and add
   all entities as the following EXAMPLE code:

```typescript
import { createEntityType, createIntegrationEntity } from './helpers';

export const [UserEntityMetadata, createUserAssignEntity] =
  createIntegrationEntity({
    resourceName: 'User',
    _class: ['User'],
    _type: createEntityType('user'), // This will generate `${INTEGRATION_NAME}_user`, but you are free to not use the createEntityType helper
    description: 'Entity description', // This will be used in the json schema
    schema: Type.Object({
      name: Type.String(),
    }),
  });
```

5. Edit the constants.ts files and replace every entity with the new entities
   created in the entities.ts file as the following EXAMPLE code:

```typescript
import { UserEntityMetadata } from './entities';

export const Entities: Record<
  | 'USER'
  StepEntityMetadata
> = {
  USER: UserEntityMetadata,
};
```

6. Now you can go into every step converter and add your create entity helper to
   `createIntegrationEntity` assign property and have full autocomplete as the
   following EXAMPLE code:

```diff
export function createUserEntity(name: string): Entity {
  return createIntegrationEntity({
    entityData: {
      source: {},
-     assign: {
+     assign: createUserAssignEntity({
        _key: `${Entities.ACCOUNT._type}|${_integrationInstanceId}`,
        name,
-     },
+     }),
    },
  });
}
```
