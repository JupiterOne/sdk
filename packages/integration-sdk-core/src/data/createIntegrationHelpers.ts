import { Static, TObject, TRef, Type } from '@sinclair/typebox';
import { StepEntityMetadata } from '../types';

interface CreateIntegrationHelpersOptions<
  IntegrationName extends string,
  SchemaKey extends string,
  ClassSchemaMap extends Record<SchemaKey, TObject>,
> {
  integrationName: IntegrationName;
  classSchemaMap: ClassSchemaMap;
}

export const createIntegrationHelpers = <
  IntegrationName extends string,
  SchemaKey extends string,
  ClassSchemaMap extends Record<SchemaKey, TObject>,
>({
  integrationName,
  classSchemaMap,
}: CreateIntegrationHelpersOptions<
  IntegrationName,
  SchemaKey,
  ClassSchemaMap
>) => {
  const createEntityType = <EntityName extends string>(
    entityName: EntityName,
  ) => `${integrationName}_${entityName}` as const;

  const createEntityMetadata = <
    ResourceName extends string,
    Class extends keyof ClassSchemaMap & string,
    EntityType extends string,
    Schema extends TObject,
  >({
    resourceName,
    _class,
    _type,
    description,
    schema,
    ...entityMetadata
  }: Omit<StepEntityMetadata, 'schema'> & {
    resourceName: ResourceName;
    _class: [Class, ...Class[]];
    _type: EntityType;
    description: string;
    schema: Schema;
  }) => {
    const classSchemaRefs = _class.map((className) =>
      Type.Ref(classSchemaMap[className]),
    ) as [TRef<ClassSchemaMap[Class]>, ...TRef<ClassSchemaMap[Class]>[]];

    const baseSchema = Type.Composite([
      Type.Object({
        _class: Type.Tuple(_class.map((className) => Type.Literal(className))),
        _type: Type.Literal(_type),
      }),
      schema,
    ]);

    const entitySchema = Type.Intersect([...classSchemaRefs, baseSchema], {
      $id: `#${_type}`,
      description: description,
    });
    type EntitySchemaType = Omit<Static<typeof entitySchema>, 'displayName'> & {
      displayName?: string | undefined;
    };

    const createEntityData = (
      entityData: Omit<EntitySchemaType, '_class' | '_type'>,
    ): EntitySchemaType => {
      return {
        ...entityData,
        _class: _class,
        _type: _type,
      } as EntitySchemaType;
    };

    const stepEntityMetadata = {
      _class,
      _type,
      resourceName,
      schema: Type.Strict(entitySchema),
      ...entityMetadata,
    } satisfies StepEntityMetadata;

    return [stepEntityMetadata, createEntityData] as const;
  };

  return { createEntityType, createEntityMetadata };
};
