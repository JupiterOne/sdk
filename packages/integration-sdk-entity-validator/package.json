{
  "name": "@jupiterone/integration-sdk-entity-validator",
  "version": "13.1.2",
  "description": "Validator for JupiterOne integration entities",
  "main": "dist/src/index.js",
  "types": "dist/src/index.d.ts",
  "repository": "git@github.com:JupiterOne/sdk.git",
  "author": "JupiterOne <dev@jupiterone.io>",
  "license": "MPL-2.0",
  "files": [
    "dist"
  ],
  "engines": {
    "node": ">=18.0.0 <21.x"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "test": "jest",
    "prebuild:dist": "rm -rf dist && mkdir dist",
    "build:dist": "tsc -p tsconfig.dist.json --declaration",
    "prepack": "npm run build:dist"
  },
  "devDependencies": {
    "@types/node": "^18.17.0",
    "ts-node": "10.9.2",
    "typescript": "5.5.2"
  },
  "dependencies": {
    "ajv": "^8.12.0",
    "ajv-formats": "^3.0.1",
    "prettier": "^3.2.5"
  }
}
