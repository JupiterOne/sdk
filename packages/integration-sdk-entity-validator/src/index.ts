export * from './validator';
export {
  unknownPropertySymbol,
  isEntityValidationError,
  type EntityValidationError,
} from './entityValidationError';
export { getValidator, getValidatorSync } from './singleton';
