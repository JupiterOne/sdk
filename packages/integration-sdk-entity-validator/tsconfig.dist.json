{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "sourceMap": true
  },
  "exclude": [
    "jest.config.js",
    "**/dist/**/*",
    "**/*.test.ts",
    "**/__tests__/**/*.ts",
    "**/__mocks__/**/*.ts"
  ]
}
