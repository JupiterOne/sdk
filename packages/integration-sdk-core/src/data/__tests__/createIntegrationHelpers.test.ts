import { Type } from '@sinclair/typebox';
import { typeboxClassSchemaMap } from '@jupiterone/data-model';
import { createIntegrationHelpers } from '../createIntegrationHelpers';
import { EntityValidator } from '@jupiterone/integration-sdk-entity-validator';
import { entitySchemas } from '@jupiterone/data-model';

describe('createIntegrationHelpers', () => {
  const { createEntityType, createEntityMetadata } = createIntegrationHelpers({
    integrationName: 'test',
    classSchemaMap: typeboxClassSchemaMap,
  });

  test('createEntityType', () => {
    expect(createEntityType('entity')).toBe('test_entity');
  });

  test('createIntegrationEntity createEntity', () => {
    const [, createEntity] = createEntityMetadata({
      resourceName: 'Entity',
      _class: ['Entity'],
      _type: 'entity',
      description: 'Entity description',
      schema: Type.Object({
        id: Type.String(),
        name: Type.String(),
      }),
    });

    expect(
      createEntity({
        id: '1',
        name: 'entity',
        _key: 'id:123456',
        displayName: 'Entity',
      }),
    ).toEqual({
      id: '1',
      name: 'entity',
      _key: 'id:123456',
      displayName: 'Entity',
      _class: ['Entity'],
      _type: 'entity',
    });
  });

  test('createIntegrationEntity schema', () => {
    const [{ schema }] = createEntityMetadata({
      resourceName: 'Entity',
      _class: ['Entity'],
      _type: 'entity',
      description: 'Entity description',
      schema: Type.Object({
        id: Type.String(),
        name: Type.String(),
        someNewProperty: Type.String(),
        thisOneIsNotRequired: Type.Optional(Type.String()),
      }),
    });

    expect(schema).toEqual({
      $id: '#entity',
      description: 'Entity description',
      allOf: [
        { $ref: '#Entity' },
        {
          properties: {
            id: { type: 'string' },
            name: { type: 'string' },
            _class: {
              type: 'array',
              items: [
                {
                  const: 'Entity',
                  type: 'string',
                },
              ],
              additionalItems: false,
              maxItems: 1,
              minItems: 1,
            },
            _type: { const: 'entity', type: 'string' },
            someNewProperty: {
              type: 'string',
            },
            thisOneIsNotRequired: {
              type: 'string',
            },
          },
          required: ['_class', '_type', 'id', 'name', 'someNewProperty'],
          type: 'object',
        },
      ],
    });
  });

  test('createIntegrationEntity schema should validate createEntity return', () => {
    const validator = new EntityValidator({
      schemas: Object.values(entitySchemas),
    });

    const [{ schema }, createEntity] = createEntityMetadata({
      resourceName: 'Type',
      _class: ['Entity'],
      _type: 'type',
      description: 'Type description',
      schema: Type.Object({
        id: Type.String(),
        name: Type.String(),
        aNewProperty: Type.Number(),
      }),
    });
    validator.addSchemas(schema);

    const { errors } = validator.validateEntity(
      createEntity({
        id: '1',
        name: 'type',
        _key: 'id:123456789',
        displayName: 'Type',
        aNewProperty: 22,
      }),
    );

    expect(errors).toEqual(null);
  });
});
