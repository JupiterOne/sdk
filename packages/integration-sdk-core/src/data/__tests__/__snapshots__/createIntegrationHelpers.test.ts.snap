// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createIntegrationHelpers createMultiClassEntityMetadata 1`] = `
[
  {
    "message": "must have required property '_key'",
    "property": "_key",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'name'",
    "property": "name",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'displayName'",
    "property": "displayName",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'category'",
    "property": "category",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'displayName'",
    "property": "displayName",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'ipv4Addresses'",
    "property": "ipv4Addresses",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'ipv6Addresses'",
    "property": "ipv6Addresses",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'macAddresses'",
    "property": "macAddresses",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'publicIpAddresses'",
    "property": "publicIpAddresses",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'privateIpAddresses'",
    "property": "privateIpAddresses",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'hostname'",
    "property": "hostname",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'fqdn'",
    "property": "fqdn",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'serial'",
    "property": "serial",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'deviceId'",
    "property": "deviceId",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'lastSeenOn'",
    "property": "lastSeenOn",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'make'",
    "property": "make",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'model'",
    "property": "model",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'osName'",
    "property": "osName",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'osType'",
    "property": "osType",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'osDetails'",
    "property": "osDetails",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'osVersion'",
    "property": "osVersion",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property '_key'",
    "property": "_key",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'name'",
    "property": "name",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'displayName'",
    "property": "displayName",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'category'",
    "property": "category",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'displayName'",
    "property": "displayName",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'ipv4Addresses'",
    "property": "ipv4Addresses",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'ipv6Addresses'",
    "property": "ipv6Addresses",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'macAddresses'",
    "property": "macAddresses",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'publicIpAddresses'",
    "property": "publicIpAddresses",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'privateIpAddresses'",
    "property": "privateIpAddresses",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'hostname'",
    "property": "hostname",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'fqdn'",
    "property": "fqdn",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'serial'",
    "property": "serial",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'deviceId'",
    "property": "deviceId",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'lastSeenOn'",
    "property": "lastSeenOn",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'make'",
    "property": "make",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'model'",
    "property": "model",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'osName'",
    "property": "osName",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'osType'",
    "property": "osType",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'osDetails'",
    "property": "osDetails",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'osVersion'",
    "property": "osVersion",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'id'",
    "property": "id",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'name'",
    "property": "name",
    "schemaId": "#multi",
    "validation": "required",
  },
  {
    "message": "must have required property 'someNewProperty'",
    "property": "someNewProperty",
    "schemaId": "#multi",
    "validation": "required",
  },
]
`;
